msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/stress/AbstractMsalUiStressTest.java

```diff
@@ -0,0 +1,264 @@
+import com.microsoft.identity.client.ui.automation.performance.DeviceMonitor;
+import com.microsoft.identity.client.ui.automation.performance.MemoryMonitor;
+import com.microsoft.identity.client.ui.automation.performance.NetworkUsageMonitor;
+import com.microsoft.identity.client.ui.automation.performance.PerformanceProfile;
+import com.microsoft.identity.client.ui.automation.rules.RulesHelper;
+import com.microsoft.identity.client.ui.automation.utils.CommonUtils;
+import com.microsoft.identity.internal.testutils.labutils.LabConstants;
+import com.microsoft.identity.internal.testutils.labutils.LabUserQuery;
+
+import org.junit.Before;
+import org.junit.rules.RuleChain;
+import org.junit.rules.Timeout;
+
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+
+public abstract class AbstractMsalUiStressTest<T, S> extends AbstractMsalUiTest {
+
+    // Sets the interval duration in seconds to which the device performance will be monitored.
+    private static final long DEVICE_MONITOR_INTERVAL = TimeUnit.SECONDS.toMillis(1);
+    private static final String DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+    private Exception executionException;
+
+    private int testsPassed = 0;
+    private int testsFailed = 0;
+
+    @Override
+    @Before
+    public void setup() {
+        super.setup();
+        executionException = null;
+
+        DeviceMonitor.setProfiler(PerformanceProfile.CPU, new CPUMonitor());
+        DeviceMonitor.setProfiler(PerformanceProfile.MEMORY, new MemoryMonitor());
+        DeviceMonitor.setProfiler(PerformanceProfile.NETWORK, new NetworkUsageMonitor());
+    }
+
+    /**
+     * Run the stress tests
+     */
+    public void run() throws Exception {
+        final T prerequisites = this.prepare();
+
+        final BlockingQueue<Runnable> blockingQueue = new ArrayBlockingQueue<>(getNumberOfThreads());
+        final RejectedExecutionHandler rejectedExecutionHandler = new ThreadPoolExecutor.CallerRunsPolicy();
+
+        final ExecutorService executorService = new ThreadPoolExecutor(
+                1,
+                getNumberOfThreads() + 1,
+                0L,
+                TimeUnit.MILLISECONDS,
+                blockingQueue,
+                rejectedExecutionHandler
+        );
+
+
+        final long startTime = System.currentTimeMillis();
+        final long timeLimit = TimeUnit.MINUTES.toMillis(getTimeLimit());
+
+        // use the execution thread pool to collect device stats
+        executorService.submit(getPerformanceStatsCollector());
+
+
+        while (System.currentTimeMillis() - startTime < timeLimit && executionException == null) {
+            executorService.submit(new Runnable() {
```

Comment:
Hm.  Rather than submitting new runnables for each of the tests, you'll get better stress-behavior if you submit a runnable that runs a series of tests.  You're looking to run 10000 tests, say, and use 10 threads, so you divide the group of 10000 into 10 groups of 100, and submit each group of 100 to 10 threads.  But since you're tests are *identical*, you could submit the workers to the test pool and just have them check whether all the tests had been run yet, run one, and then report.

Reply:
So basically, rather than trying to pre-divide the work, just use a series of atomics to count to completion, and then make sure that the number or requested tests actually get run as opposed to some test-software exception causing premature termination.

Reply:
I may not be able to predict how many tests we can run, and I would like to run as many tests as possible. So, I've done a similar implementation by submitting only `n` number of threads and allowing each thread to run a series of tests while having atomics to count the number of tests that have been run (evaluating both failure and success)